### ISSUE TITLE
Resilver restarting?

### ISSUE THREAD

**Issue by chrisrd** [zfsonlinux/zfs#840]
I used `zpool replace ${pool} ${guid} ${dev}` to replace one dead and one dying disk, which started a resilver. I've been monitoring the progress and have seen it apparently restart, e.g.:

```
while true; do ts=$(date +"%F-%T"); echo -e "$ts \c"; zpool status pool2 | egrep scanned; sleep 30; done
...
2012-07-19-13:55:09     1.59T scanned out of 29.9T at 286M/s, 28h49m to go
2012-07-19-13:55:39     1.60T scanned out of 29.9T at 286M/s, 28h48m to go
2012-07-19-13:56:09     53.7M scanned out of 29.9T at 1.99M/s, (scan is slow, no estimated time)
2012-07-19-13:56:39     89.4M scanned out of 29.9T at 1.57M/s, (scan is slow, no estimated time)
...
```

In conjunction with the apparent restart there's a disk error reported in the `kern.log`:

```
Jul 19 13:55:35 b5 kernel: [81414.117130] mpt2sas0: log_info(0x31080000): originator(PL), code(0x08), sub_code(0x0000)
Jul 19 13:55:35 b5 kernel: [81414.117194] sd 0:0:57:0: [sdbc] Unhandled sense code
Jul 19 13:55:35 b5 kernel: [81414.117223] sd 0:0:57:0: [sdbc]  Result: hostbyte=invalid driverbyte=DRIVER_SENSE
Jul 19 13:55:35 b5 kernel: [81414.117276] sd 0:0:57:0: [sdbc]  Sense Key : Medium Error [current] 
Jul 19 13:55:35 b5 kernel: [81414.117311] Info fld=0x5066f2a4
Jul 19 13:55:35 b5 kernel: [81414.117333] sd 0:0:57:0: [sdbc]  Add. Sense: Unrecovered read error
Jul 19 13:55:35 b5 kernel: [81414.117366] sd 0:0:57:0: [sdbc] CDB: Read(10): 28 00 50 66 f2 6c 00 00 39 00
Jul 19 13:55:35 b5 kernel: [81414.117422] end_request: critical target error, dev sdbc, sector 1348924012
Jul 19 13:55:35 b5 kernel: [81414.117455] ZFS: zio error=121 type=1 offset=688647362560 size=29184 flags=20668 delay=7000
```

...where `sdbc` is one of the drives being replaced.

This has happened at least 4 or 5 times since the replaces were originally done some 24 hours ago, and the 3 times I've actually watched it happen have all been with errors on `sdbc` (at different sectors).  

On the other hand, there've been other disk errors, including (I think) on `sdbc`, that haven't caused the resilver to apparently restart.

Is this normal, and/or can I expect the resilver to actually finish, or is it stuck in a loop?

Oh, I've just realised I have a few dumps of the complete `zfs status` taken during all this, which show that as far as the `in progress since...` message is concerned the resilver really is restarting. E.g. from this morning vs now:

```
scan: resilver in progress since Thu Jul 19 05:08:45 2012
scan: resilver in progress since Thu Jul 19 14:47:10 2012
```

The complete current `zpool status`, with the /dev/mapper devices annotated with their matching /dev device:

```
  pool: pool2
 state: DEGRADED
status: One or more devices is currently being resilvered.  The pool will
	continue to function, possibly in a degraded state.
action: Wait for the resilver to complete.
 scan: resilver in progress since Thu Jul 19 14:47:10 2012
    653G scanned out of 29.9T at 392M/s, 21h43m to go
    39.4G resilvered, 2.13% done
config:

	NAME                                        STATE     READ WRITE CKSUM
	pool2                                  DEGRADED     0     0     0
	  raidz2-0                                  ONLINE       0     0     0
	    sdbk  Hitachi_HDS5C30_ML0220F30E2DMD    ONLINE       0     0     0
	    sdbi  Hitachi_HDS5C30_ML0221F30B7ZND    ONLINE       0     0     0
	    sdac  Hitachi_HDS7230_MN1240F33XBYYD    ONLINE       0     0     0
	    sdad  Hitachi_HDS7230_MN1240FA03E4HD    ONLINE       0     0     0
	    sdab  Hitachi_HDS7230_MN1240FA0HUZAD    ONLINE       0     0     0
	    sdaa  Hitachi_HDS7230_MN1240FA0JXU2D    ONLINE       0     0     0
	    sdag  Hitachi_HDS7230_MN1240FA0K07YD    ONLINE       0     0     0
	    sdaf  Hitachi_HDS7230_MN1240FA0KP7YD    ONLINE       0     0     0
	    sdah  Hitachi_HDS7230_MN1240FA0LA36D    ONLINE       0     0     0
	    sdae  Hitachi_HDS7230_MN1240FA0LMP9D    ONLINE       0     0     0
	    sdz   Hitachi_HDS7230_MN1270FA0SV8ED    ONLINE       0     0     0
	  raidz2-1                                  ONLINE       0     0     0
	    sdl   ST2000DL003-9VT_5YD1V6ZL          ONLINE       0     0     0
	    sdaj  ST2000DL003-9VT_5YD1W04H          ONLINE       0     0     0
	    sdn   ST2000DL003-9VT_5YD1XEDS          ONLINE       0     0     0
	    sdi   ST2000DL003-9VT_5YD1XF0K          ONLINE       0     0     0
	    sdak  ST2000DL003-9VT_5YD1XQ15          ONLINE       0     0     0
	    sdo   ST2000DL003-9VT_5YD1Z9BZ          ONLINE       0     0     0
	    sdbf  WDC_WD20EARX-00_WD-WMAZA5589913   ONLINE       0     0  816K
	    sdai  ST2000DL003-9VT_5YD563JW          ONLINE       0     0     0
	    sds   ST2000DL003-9VT_5YD56MEZ          ONLINE       0     0     0
	    sdy   ST2000DL003-9VT_5YD5FWQ6          ONLINE       0     0     0
	    sdbs  WDC_WD20EARX-00_WD-WMAZA5567696   ONLINE       0     0     0
	  raidz2-2                             DEGRADED     3     0     0
	    sdbe  WDC_WD20EADS-00_WD-WCAVY1129416   ONLINE     110     0     1  (resilvering)
	    sdaz  WDC_WD20EADS-00_WD-WCAVY1162803   ONLINE     120     0     2
	          replacing-2                       ONLINE       0     0    39
	      sdbc  WDC_WD20EADS-00_WD-WCAVY1184604 ONLINE      70     0     0  (resilvering)
	      sdbj  WDC_WD20EARX-00_WD-WMAZA5603921 ONLINE       0     0     0  (resilvering)
	    sdbh  WDC_WD20EADS-00_WD-WCAVY1188798   ONLINE     158     0     0
	          replacing-4                       UNAVAIL      0     0     0
	            17433600781902467301            UNAVAIL      0     0     0  was /dev/mapper/WDC_WD20EADS-00_WD-WCAVY1188828
	      sdba  Hitachi_HDS7230_MN1240FA0LHJLD  ONLINE       0     0     0  (resilvering)
	    sdbr  WDC_WD20EADS-00_WD-WCAVY1188933   ONLINE     153     0     2
	    sdbt  WDC_WD20EADS-00_WD-WCAVY7404957   ONLINE       0     0     0
	    sdbg  WDC_WD20EARX-00_WD-WCAZA9871198   ONLINE       0     0     0
	    sdbq  WDC_WD20EARX-00_WD-WMAZA5346235   ONLINE       0     0     0
	    sdbp  WDC_WD20EARX-00_WD-WMAZA5534735   ONLINE       0     0     0
	    sdbl  WDC_WD20EARX-00_WD-WMAZA5561260   ONLINE       0     0     0

errors: No known data errors
```

**Comment by thisisnotmyrealname** [comment]
I really hate bumping old tickets, but I'm not sure what to do. 

I have the exact same issue running 0.6.5-304_gf74b821

I had a problem where a failing drive was still present in the machine, and repeatedly caused the pool to restart resilvering due to possibly this problem? I had my resilver restart about 5 times before I finally found this, offlined the bad device and removed it from the chassis.

My setup is a supermicro enclosure with SATA disks behind SAS HBA/Expanders.
`

**Comment by tobias-k** [comment]
Hey, 
I also hate bumping.. 
In my case 4 TB SAS NL in Supermicro Expansion.. :(

`[19432.214251] 3w-sas: scsi0: AEN: INFO (0x04:0x0055): Battery charging started:.
[20022.719713] 3w-sas: scsi0: AEN: INFO (0x04:0x0056): Battery charging completed:.
[52097.645349] init: smbd main process (4243) killed by TERM signal
[70169.668106] mpt2sas0: log_info(0x31120303): originator(PL), code(0x12), sub_code(0x0303)
[70169.668115] sd 1:0:26:0: [sdaa] FAILED Result: hostbyte=DID_SOFT_ERROR driverbyte=DRIVER_OK
[70169.668129] sd 1:0:26:0: [sdaa] CDB:
[70169.668133] Write(16): 8a 00 00 00 00 01 95 ab 89 29 00 00 00 06 00 00
[70169.668147] blk_update_request: I/O error, dev sdaa, sector 6806014249
[138268.929611] 3w-sas: scsi0: AEN: INFO (0x04:0x0055): Battery charging started:.
[138560.253620] init: smbd main process (22177) killed by TERM signal
[138924.490649] 3w-sas: scsi0: AEN: INFO (0x04:0x0056): Battery charging completed:.
[225036.473711] init: smbd main process (23540) killed by TERM signal
[225248.031596] mpt2sas0: log_info(0x31120303): originator(PL), code(0x12), sub_code(0x0303)
[225248.031623] sd 1:0:26:0: [sdaa] FAILED Result: hostbyte=DID_SOFT_ERROR driverbyte=DRIVER_OK
[225248.031626] sd 1:0:26:0: [sdaa] CDB:
[225248.031629] Write(16): 8a 00 00 00 00 00 aa 36 c4 11 00 00 00 06 00 00
[225248.031655] blk_update_request: I/O error, dev sdaa, sector 2855715857
[257130.648854] 3w-sas: scsi0: AEN: INFO (0x04:0x0055): Battery charging started:.
[257861.274410] 3w-sas: scsi0: AEN: INFO (0x04:0x0056): Battery charging completed:.
[311488.671949] init: smbd main process (21718) killed by TERM signal
[331118.157025] mpt2sas0: log_info(0x31120303): originator(PL), code(0x12), sub_code(0x0303)
[331118.157064] sd 1:0:26:0: [sdaa] FAILED Result: hostbyte=DID_SOFT_ERROR driverbyte=DRIVER_OK
[331118.157071] sd 1:0:26:0: [sdaa] CDB:
[331118.157073] Write(16): 8a 00 00 00 00 00 7b 65 fe 97 00 00 00 06 00 00
[331118.157083] blk_update_request: I/O error, dev sdaa, sector 2070281879
[376032.400281] 3w-sas: scsi0: AEN: INFO (0x04:0x0055): Battery charging started:.
[376753.017234] 3w-sas: scsi0: AEN: INFO (0x04:0x0056): Battery charging completed:.
[397969.245328] init: smbd main process (20776) killed by TERM signal
[407858.723501] mpt2sas0: log_info(0x31120303): originator(PL), code(0x12), sub_code(0x0303)
[407858.723507] sd 1:0:26:0: [sdaa] FAILED Result: hostbyte=DID_SOFT_ERROR driverbyte=DRIVER_OK
[407858.723510] sd 1:0:26:0: [sdaa] CDB:
[407858.723512] Write(16): 8a 00 00 00 00 00 24 c5 9f bb 00 00 00 12 00 00
[407858.723525] blk_update_request: I/O error, dev sdaa, sector 616931259
[448206.673498] mpt2sas0: log_info(0x31120303): originator(PL), code(0x12), sub_code(0x0303)
[448206.673505] sd 1:0:26:0: [sdaa] FAILED Result: hostbyte=DID_SOFT_ERROR driverbyte=DRIVER_OK
[448206.673508] sd 1:0:26:0: [sdaa] CDB:
[448206.673510] Write(16): 8a 00 00 00 00 00 0c 70 13 00 00 00 00 21 00 00
[448206.673526] blk_update_request: I/O error, dev sdaa, sector 208671488
[474471.029891] mpt2sas0: log_info(0x31120303): originator(PL), code(0x12), sub_code(0x0303)
[474471.029909] sd 1:0:26:0: [sdaa] FAILED Result: hostbyte=DID_SOFT_ERROR driverbyte=DRIVER_OK
[474471.029912] sd 1:0:26:0: [sdaa] CDB:
[474471.029914] Write(16): 8a 00 00 00 00 00 65 89 d3 90 00 00 00 60 00 00
[474471.029923] blk_update_request: I/O error, dev sdaa, sector 1703531408
[484441.953357] init: smbd main process (4546) killed by TERM signal
[494924.144510] 3w-sas: scsi0: AEN: INFO (0x04:0x0055): Battery charging started:.
[495634.752488] 3w-sas: scsi0: AEN: INFO (0x04:0x0056): Battery charging completed:.
[570913.751570] init: smbd main process (25086) killed by TERM signal
[613805.880539] 3w-sas: scsi0: AEN: INFO (0x04:0x0055): Battery charging started:.
[614436.420071] 3w-sas: scsi0: AEN: INFO (0x04:0x0056): Battery charging completed:.
[646135.203418] mpt2sas0: log_info(0x31120303): originator(PL), code(0x12), sub_code(0x0303)
[646135.203504] sd 1:0:26:0: [sdaa] FAILED Result: hostbyte=DID_SOFT_ERROR driverbyte=DRIVER_OK
[646135.203508] sd 1:0:26:0: [sdaa] CDB:
[646135.203509] Write(16): 8a 00 00 00 00 01 9f 82 e9 ff 00 00 00 18 00 00
[646135.203520] blk_update_request: I/O error, dev sdaa, sector 6971124223
[647217.608785] mpt2sas0: log_info(0x31120303): originator(PL), code(0x12), sub_code(0x0303)
[647217.608791] sd 1:0:43:0: [sdar] FAILED Result: hostbyte=DID_SOFT_ERROR driverbyte=DRIVER_OK
[647217.608794] sd 1:0:43:0: [sdar] CDB:
[647217.608796] Write(16): 8a 00 00 00 00 00 76 68 47 ec 00 00 00 08 00 00
[647217.608812] blk_update_request: I/O error, dev sdar, sector 1986545644
[657414.346777] init: smbd main process (18966) killed by TERM signal
[690535.415451] mpt2sas0: log_info(0x31120303): originator(PL), code(0x12), sub_code(0x0303)
[690535.415454] sd 1:0:26:0: [sdaa] FAILED Result: hostbyte=DID_SOFT_ERROR driverbyte=DRIVER_OK
[690535.415458] sd 1:0:26:0: [sdaa] CDB:
[690535.415460] Write(16): 8a 00 00 00 00 00 a2 d9 dc 57 00 00 00 06 00 00
[690535.415474] blk_update_request: I/O error, dev sdaa, sector 2732186711
[732622.569088] 3w-sas: scsi0: AEN: INFO (0x04:0x0055): Battery charging started:.
[733328.174676] 3w-sas: scsi0: AEN: INFO (0x04:0x0056): Battery charging completed:.
[743843.444937] init: smbd main process (11716) killed by TERM signal
[819092.268208] mpt2sas0: log_info(0x31120303): originator(PL), code(0x12), sub_code(0x0303)
[819092.268214] sd 1:0:26:0: [sdaa] FAILED Result: hostbyte=DID_SOFT_ERROR driverbyte=DRIVER_OK
[819092.268216] sd 1:0:26:0: [sdaa] CDB:
[819092.268218] Write(16): 8a 00 00 00 00 00 11 b1 8e f9 00 00 00 12 00 00
[819092.268230] blk_update_request: I/O error, dev sdaa, sector 296849145
[830323.662533] init: smbd main process (19837) killed by TERM signal
[851494.300817] 3w-sas: scsi0: AEN: INFO (0x04:0x0055): Battery charging started:.
[852144.857705] 3w-sas: scsi0: AEN: INFO (0x04:0x0056): Battery charging completed:.
`
`root@fshh7:~# zpool status
  pool: fshh7
 state: DEGRADED
status: One or more devices is currently being resilvered.  The pool will
        continue to function, possibly in a degraded state.
action: Wait for the resilver to complete.
  scan: resilver in progress since Sun Jun 19 22:54:38 2016
    14.1T scanned out of 99.2T at 408M/s, 60h43m to go
    314G resilvered, 14.17% done
config:

        NAME                          STATE     READ WRITE CKSUM
        fshh7                         DEGRADED     0     0     0
          raidz2-0                    ONLINE       0     0     0
            sdau                      ONLINE       0     0     0
            sdav                      ONLINE       0     0     0
            sdaw                      ONLINE       0     0     0
            sday                      ONLINE       0     0     0
            sdax                      ONLINE       0     0     0
            sdaz                      ONLINE       0     0     0
            sdba                      ONLINE       0     0     0
            sdbb                      ONLINE       0     0     0
            sdbc                      ONLINE       0     0     0
            sdbd                      ONLINE       0     0     0
            sdbe                      ONLINE       0     0     0
            sdbf                      ONLINE       0     0     0
            sdbg                      ONLINE       0     0     0
            sdbh                      ONLINE       0     0     0
            sdbi                      ONLINE       0     0     0
            sdbj                      ONLINE       0     0     0
            sdbk                      ONLINE       0     0     0
            sdbl                      ONLINE       0     0     0
            sdbm                      ONLINE       0     0     0
            sdbn                      ONLINE       0     0     0
            sdbo                      ONLINE       0     0     0
            sdbp                      ONLINE       0     0     0
          raidz2-2                    DEGRADED     0     0     0
            sdb                       ONLINE       0     0     0
            sdc                       ONLINE       0     0     0
            sdd                       ONLINE       0     0     0
            sde                       ONLINE       0     0     0
            sdf                       ONLINE       0     0     0
            sdg                       ONLINE       0     0     0
            sdh                       ONLINE       0     0     0
            sdi                       ONLINE       0     0     0
            sdj                       ONLINE       0     0     0
            sdk                       ONLINE       0     0     0
            sdl                       ONLINE       0     0     0
            sdm                       ONLINE       0     0     0
            sdn                       ONLINE       0     0     0
            sdo                       ONLINE       0     0     0
            sdp                       ONLINE       0     0     0
            sdq                       ONLINE       0     0     0
            sdr                       ONLINE       0     0     0
            sds                       ONLINE       0     0     0
            sdt                       ONLINE       0     0     0
            sdu                       ONLINE       0     0     0
            sdv                       ONLINE       0     0     0
            sdw                       ONLINE       0     0     0
            sdx                       ONLINE       0     0     0
            sdy                       ONLINE       0     0     0
            sdz                       ONLINE       0     0     0
            replacing-25              DEGRADED     0     0     0
              6262255769558981431     FAULTED      0     0     0  was /dev/sdaa1
              scsi-35000c50083acd193  ONLINE       0     0     0  (resilvering)
            sdab                      ONLINE       0     0     0
            sdac                      ONLINE       0     0     0
            sdad                      ONLINE       0     0     0
            sdae                      ONLINE       0     0     0
            sdaf                      ONLINE       0     0     0
            sdag                      ONLINE       0     0     0
            sdah                      ONLINE       0     0     0
            sdai                      ONLINE       0     0     0
            sdaj                      ONLINE       0     0     0
            sdak                      ONLINE       0     0     0
            sdal                      ONLINE       0     0     0
            sdam                      ONLINE       0     0     0
            sdan                      ONLINE       0     0     0
            sdao                      ONLINE       0     0     0
            sdap                      ONLINE       0     0     0
            sdaq                      ONLINE       0     0     0
            sdar                      ONLINE       0     0     0
            sdas                      ONLINE       0     0     0
            sdat                      ONLINE       0     0     0
        logs
          mirror-1                    ONLINE       0     0     0
            sdbq1                     ONLINE       0     0     0
            sdbr1                     ONLINE       0     0     0
        cache
          sdbq2                       ONLINE       0     0     0
          sdbr2                       ONLINE       0     0     0
`

**Comment by rincebrain** [comment]
There's probably a bug to be found with this, but as a mitigation for either of you, do you have TLER configured on the failing drives? (e.g. smartctl -l scterc,RNUM,WNUM [drive] - RNUM/WNUM are tenths of a second to attempt read/write before giving up, while 0 will generally cause it to not give up...which, if it takes too long, can be when the various other parts of the chain decide to reset it forcibly)

I'd suggest trying it if you can, because presuming the implementation isn't buggy, most parts of IO stacks respond better to drives replying after X seconds that they've had an error than to drives which stop responding entirely until you reset the device.

**Comment by tobias-k** [comment]
Hey, thx for reply. 

I can test it, Which Values for scterc would you recommend?
I will set them asap and see if the rebuild will finaly get trough.. 

Thank you!

**Comment by raxxy** [comment]
Tobias, I'm also suffering this issue at the moment.

I've been talking to the blokes on IRC and was given values of 70,70 to try (smartctl -l scterc,70,70)

At this stage i'm on my 5th resilver restart, I'll report back on array type and failure/resolution on this setting (I'm applying it to all disks however) in the near future..

**Comment by tobias-k** [comment]
Thank you. 
So I'll set 80,80 and give feedback. Maybe you can skip at least one resilvering. 
60 hours to go.. in my case.. 
Fingers crossed..

**Comment by tobias-k** [comment]
Ok. Again restarted. :(
Next try 150,150

**Comment by rincebrain** [comment]
Changing the values to different numbers shouldn't affect whether it helps or not, unfortunately.

**Comment by tobias-k** [comment]
So @rincebrain you would suggest value 0 anyway?

Thank you.

**Comment by rincebrain** [comment]
No, I still claim having it set to nonzero values in a RAID array is a good idea, but if one set of nonzero values didn't workaround this issue, a different set probably won't either.

**Comment by raxxy** [comment]
Reset again for me, setting this on all disks had no effect.

Might remove the new drive and copy as much data as I can off before catastrophic failure and just rebuild, hopefully removing the new replacement drive cancells the resilver..

**Comment by chrisrd** [comment]
Original problem reporter here...

If your pool is raid-z2 (e.g. @tobias-k), I'd suggest removing the faulty drive from the pool to stop it producing errors that cause the resilver to restart. You still have a remaining parity drive so your data is still safe and this allows you to get back to 2 x parity as quickly as possible.

If you have a faulty drive on raid-z1 you can either remove the faulty drive to get the resilver to complete as quickly as possible (your data is still all there, but you may lose some blocks if another drive has problems during the resilver), or, if you have another spare drive (i.e. in addition to the one you're trying to resilver to), do as I did per my "TLDR" post above and stop the resilvering (e.g. export the pool), then "dd" the faulty drive to the second spare (don't worry too much about occasional errors),  remove the faulty drive, then bring the pool back up with the new copied drive, and let the resilver complete. This second approach is a bit safer then simply removing the faulty drive because if another drive has problems during the subsequent resilver it will hopefully be covered by the new copied drive - assuming the new problem is on different blocks to where the original faulty drive was having problems.

**Comment by raxxy** [comment]
Removed old disk. Resilver completed, did NOT remove the offlined disk (still stuck there) reboots still causing a resilver restart even though it does complete - never removing the offlined disk.

I have no more data errors (i had 5 files last time, now showing as zero).

I'm just going to complete moving data off array and blow it away.

**Comment by chrisrd** [comment]
@raxxy Can you clarify what you mean by "removed old disk", then "did NOT remove the offlined disk"  please? If you mean the old failing disk is still in the machine, and a reboot causes another resilver to start again, what happens if you let the resilver complete and remove the failing disk from the machine and reboot?

**Comment by rincebrain** [comment]
I believe he means that it claims to be done resilvering/scrubbing, but the device is still marked REPLACING {offline disk, new disk}, and kicks off a resilver anew every reboot.

**Comment by raxxy** [comment]
Rincebrain is correct, sorry I was in a rush to leave for work, clearly I should proofread more often.. ;)

Dead disk + new replacement disk -> IO errors triggering resilver re-start loop
Dead disk physically removed + new replacement disk in -> resilver completes but does not ever remove the disk now physically removed

Any event to the zfs pool (export/import, reboot) will cause the resilver to re-start, and likely complete but never remove the disk (This is the 30th resilver all up, the first ones were doing the same thing - put disk back in to see if it would fix it - removed disk again so I can get some speed to move the data.)

**Comment by chrisrd** [comment]
@raxxy OK, I don't know what's going on there. Does your "zpool events -v poolname" have anything interesting in it relating to the scrubs? If you're happy compiling your own ZoL you might try adding in this patch which provides a little more "zpool events" information on the scrub completion: chrisrd/zfs@92da32e

It's also worth bringing up the pool without /etc/zfs/zpool.cache. Something like: export the pool, remove /etc/zfs/zpool.cache, and import again (or reboot and import).

**Comment by behlendorf** [comment]
Commit d14fa5dba1ad0e68e803435ac48ec1ea78121699 which was merged to master may have resolved this.  Has anyone tested with this change?

**Comment by thisisnotmyrealname** [comment]
Not fixed, Still getting restarts with drives that have seek errors.

filename:       /lib/modules/4.4.0-36-generic/kernel/zfs/zfs/zfs.ko
version:        0.6.5.6-0ubuntu10
license:        CDDL
author:         OpenZFS on Linux
description:    ZFS
srcversion:     BAF128A524F0974CFA37071
depends:        spl,znvpair,zunicode,zcommon,zavl
vermagic:       4.4.0-36-generic SMP mod_unload modversions

[Fri Sep  2 11:31:28 2016] blk_update_request: critical medium error, dev sdz, sector 4106662640
[Fri Sep  2 11:31:33 2016] sd 2:0:20:0: device_block, handle(0x001e)
[Fri Sep  2 11:31:33 2016] mpt2sas_cm2: log_info(0x31110610): originator(PL), code(0x11), sub_code(0x0610)
[Fri Sep  2 11:31:33 2016] mpt2sas_cm2: log_info(0x31110610): originator(PL), code(0x11), sub_code(0x0610)
[Fri Sep  2 11:31:33 2016] sd 2:0:20:0: device_unblock and setting to running, handle(0x001e)
[Fri Sep  2 11:31:34 2016] sd 2:0:20:0: [sdz] tag#1 CDB: Read(16) 88 00 00 00 00 00 f4 c6 b3 70 00 00 00 20 00 00
[Fri Sep  2 11:31:34 2016] mpt2sas_cm2:         sas_address(0x5003048000d25b63), phy(35)
[Fri Sep  2 11:31:34 2016] mpt2sas_cm2:         enclosure_logical_id(0x5003048000d25b7f),slot(23)
[Fri Sep  2 11:31:34 2016] mpt2sas_cm2:         handle(0x001e), ioc_status(success)(0x0000), smid(162)
[Fri Sep  2 11:31:34 2016] mpt2sas_cm2:         request_len(16384), underflow(16384), resid(16384)
[Fri Sep  2 11:31:34 2016] mpt2sas_cm2:         tag(65535), transfer_count(0), sc->result(0x00000000)
[Fri Sep  2 11:31:34 2016] mpt2sas_cm2:         scsi_status(check condition)(0x02), scsi_state(autosense valid )(0x01)
[Fri Sep  2 11:31:34 2016] mpt2sas_cm2:         [sense_key,asc,ascq]: [0x06,0x29,0x00], count(18)

blah blah blah

**Comment by rincebrain** [comment]
@thisisnotmyrealname That's not surprising, none of the 0.6.5.X releases have the commit that @behlendorf mentioned above might fix your issue.

You'd need to either build from git master or just apply that particular commit to 0.6.5.X and build, and see if the problem persists for you.

**Comment by thisisnotmyrealname** [comment]
Apologies on the previous report. Seems Ubuntu snuck the stock ZFS kernel module in on me.

filename:       /lib/modules/4.4.0-36-generic/extra/zfs/zfs.ko
version:        0.6.5-437_gd02ca37
license:        CDDL
author:         OpenZFS on Linux
description:    ZFS
srcversion:     E4190FDAA7A1780753688F9
depends:        spl,znvpair,zunicode,zcommon,zavl
vermagic:       4.4.0-36-generic SMP mod_unload modversions

I checked out the latest git against what I have compiled, and only saw changes to zfs test suite.

status: One or more devices is currently being resilvered.  The pool will
        continue to function, possibly in a degraded state.
action: Wait for the resilver to complete.
  scan: resilver in progress since Fri Sep  9 01:15:26 2016
        39.2G scanned out of 84.6T at 5.45M/s, (scan is slow, no estimated time)
        1.52G resilvered, 0.05% done

I started a resilver 7 days ago, and it's been restarting since then. Is there any other information I can provide that is useful/will help?

**Comment by TyberiusPrime** [comment]
Can confirm having had a resilver-restarts-over-and-over again issue that disappeared when switching to thursday's zfs-daily ppa. My pool is whole again.

**Comment by thegreatgazoo** [comment]
It may have something to do with #5970.

**Comment by muff1nman** [comment]
I think I have run into this same issue, but I'm not entirely sure..

I have a raidz2 pool that I had setup with misaligned partitions. I wanted to fix this, so tried offlining out drives one by one, clearing the partition table and then replacing them but instead using the full drive. However, after replacing the drives the status for the resilver process seems to restart and never progresses. For one example, zpool status gave me:

```
action: Wait for the resilver to complete.
  scan: resilver in progress since Thu Jan 11 18:19:54 2018
        252M scanned out of 6.99T at 12.6M/s, 161h24m to go
        42.3M resilvered, 0.00% done
```
And a couple seconds later I got:
```

action: Wait for the resilver to complete.
  scan: resilver in progress since Thu Jan 11 18:20:15 2018
        227M scanned out of 6.99T at 13.4M/s, 152h27m to go
        37.6M resilvered, 0.00% done
```

Unsure what to do and noticing that it seemed to always cycle when it hit a certain point, I rebooted the server, at which point the resilver started progressing and got past the sticking point. This has happened on both of the drives I have tried repartitioning in my pool. Attached are the logs of the attempt for the second drive [zpool_replace.log.redacted.gz](https://github.com/zfsonlinux/zfs/files/1626903/zpool_replace.log.redacted.gz) (also pasted at [https://paste.fedoraproject.org/paste/HbUQxhWB6L7j59FNLCFzHA](https://paste.fedoraproject.org/paste/HbUQxhWB6L7j59FNLCFzHA)). I did run a scrub after doing the first drive (which was successful) before going on to the second drive.

The one clue I did find was this in the journal logs around the time of a resilver restart:

```
Jan 11 18:18:54 ZFSHOST zed[8820]: eid=488 class=config_sync pool_guid=0x89A3B4EB7BFA7DBF
Jan 11 18:18:55 ZFSHOST zed[9045]: eid=489 class=config_sync pool_guid=0x89A3B4EB7BFA7DBF
Jan 11 18:18:55 ZFSHOST zed[9047]: eid=490 class=config_sync pool_guid=0x89A3B4EB7BFA7DBF
Jan 11 18:18:56 ZFSHOST zed[9270]: eid=491 class=config_sync pool_guid=0x89A3B4EB7BFA7DBF
Jan 11 18:18:56 ZFSHOST zed[9280]: eid=492 class=config_sync pool_guid=0x89A3B4EB7BFA7DBF
Jan 11 18:18:56 ZFSHOST kernel: Alternate GPT is invalid, using primary GPT.
Jan 11 18:18:56 ZFSHOST kernel:  sda:
Jan 11 18:18:56 ZFSHOST kernel:  sda: sda1 sda9
Jan 11 18:18:56 ZFSHOST zed[9382]: eid=493 class=vdev_attach pool_guid=0x89A3B4EB7BFA7DBF vdev_path=/dev/disk/by
Jan 11 18:18:57 ZFSHOST zed[9539]: eid=494 class=resilver_start pool_guid=0x89A3B4EB7BFA7DBF
Jan 11 18:18:57 ZFSHOST zed[9545]: eid=495 class=history_event pool_guid=0x89A3B4EB7BFA7DBF
Jan 11 18:19:01 ZFSHOST zed[11291]: eid=496 class=config_sync pool_guid=0x89A3B4EB7BFA7DBF
Jan 11 18:19:06 ZFSHOST zed[13682]: eid=497 class=history_event pool_guid=0x89A3B4EB7BFA7DBF
Jan 11 18:19:20 ZFSHOST zed[18764]: eid=498 class=config_sync pool_guid=0x89A3B4EB7BFA7DBF
Jan 11 18:19:20 ZFSHOST zed[18773]: eid=499 class=config_sync pool_guid=0x89A3B4EB7BFA7DBF
Jan 11 18:19:35 ZFSHOST zed[26064]: eid=500 class=config_sync pool_guid=0x89A3B4EB7BFA7DBF
Jan 11 18:19:35 ZFSHOST zed[26071]: eid=501 class=config_sync pool_guid=0x89A3B4EB7BFA7DBF
Jan 11 18:19:39 ZFSHOST zed[27625]: eid=502 class=history_event pool_guid=0x89A3B4EB7BFA7DBF
Jan 11 18:19:39 ZFSHOST zed[27627]: eid=503 class=resilver_start pool_guid=0x89A3B4EB7BFA7DBF
Jan 11 18:19:39 ZFSHOST zed[27637]: eid=504 class=history_event pool_guid=0x89A3B4EB7BFA7DBF
Jan 11 18:19:50 ZFSHOST zed[32020]: eid=505 class=config_sync pool_guid=0x89A3B4EB7BFA7DBF
Jan 11 18:19:50 ZFSHOST zed[32025]: eid=506 class=config_sync pool_guid=0x89A3B4EB7BFA7DBF
Jan 11 18:19:54 ZFSHOST zed[1317]: eid=507 class=history_event pool_guid=0x89A3B4EB7BFA7DBF
Jan 11 18:19:54 ZFSHOST zed[1323]: eid=508 class=resilver_start pool_guid=0x89A3B4EB7BFA7DBF
Jan 11 18:19:54 ZFSHOST zed[1327]: eid=509 class=history_event pool_guid=0x89A3B4EB7BFA7DBF
```

**Comment by rincebrain** [comment]
If I had to guess, it's getting confused by the fact that you didn't tell it you were clobbering the old device, and might be stuck trying to relabel/import the "old" drive.

I'm reasonably confident what's happening to you is something is triggering a relabeling of the drive and, when the drive vanishes and reappears, thereby triggering the resilver to restart.

I'm less confident about _what_ is triggering it, but I'd guess that it's eventually noticing the part1 "old" drive reappearing when it relabels, attempting to online it, and in doing so, clobbering the resilver.

I'd use zpool offline to explicitly offline the "old" device, which should probably convince it to stop trying this; if not, then making sure autoreplace=off on the pool should; if that doesn't, then that's another bug, I would think.

(Which ZoL version, distro+kernel version, etc?)

**Comment by muff1nman** [comment]
I did do a zpool offline as per the log history I attached. Also autoreplace is set to off:
```
# zpool get all | grep autoreplace
tank  autoreplace                    off                            default
```
This is version 0.7.5-1 on Centos 7 kernel 3.10.0-693.11.6.el7.x86_64.

**Comment by StefanCohen** [comment]
I'm on version 0.7.3-1 and I encountered the same issue yesterday after a "zpool replace <pool> <olddrive> <newdrive>". I noticed that the resilvering restarted repeatedly every ten seconds with a ton of these in the zfs history:
```
2018-02-04.00:58:42 [txg:7268183] scan setup func=2 mintxg=3 maxtxg=7264178
2018-02-04.00:59:08 [txg:7268189] scan aborted, restarting errors=0
```
I searched and found the following issue: [6613](https://github.com/zfsonlinux/zfs/issues/6613). It was mentioned that it could be related to zed so i stopped zfs-zed and it resilvered in about an hour after that. I could reproduce the problem on another disk so this might in fact be a bug related to ZED.

**Comment by as-com** [comment]
I'm using archzfs-git (probably a very recent version) and I was having issues replacing a drive, with these logs in the journal:
```
Mar 11 22:43:47 hubble zed[19428]: eid=9385 class=config_sync pool_guid=0x38BF9374E2B7FD2D
Mar 11 22:43:50 hubble zed[19788]: eid=9386 class=config_sync pool_guid=0x38BF9374E2B7FD2D
Mar 11 22:43:51 hubble zed[19867]: eid=9387 class=vdev.bad_ashift pool_guid=0x38BF9374E2B7FD2D vdev_path=/dev/disk/by-id/wwn-0x50014ee262071313-part3
Mar 11 22:43:51 hubble zed[19892]: eid=9388 class=vdev.bad_ashift pool_guid=0x38BF9374E2B7FD2D vdev_path=/dev/disk/by-id/wwn-0x50014ee20cb16a07-part3
Mar 11 22:43:52 hubble zed[19912]: eid=9389 class=history_event pool_guid=0x38BF9374E2B7FD2D
Mar 11 22:43:52 hubble zed[19916]: eid=9390 class=resilver_start pool_guid=0x38BF9374E2B7FD2D
Mar 11 22:43:52 hubble zed[19918]: eid=9391 class=history_event pool_guid=0x38BF9374E2B7FD2D
Mar 11 22:44:02 hubble zed[20670]: eid=9392 class=config_sync pool_guid=0x38BF9374E2B7FD2D
Mar 11 22:44:06 hubble zed[20941]: eid=9393 class=config_sync pool_guid=0x38BF9374E2B7FD2D
Mar 11 22:44:06 hubble zed[21032]: eid=9394 class=vdev.bad_ashift pool_guid=0x38BF9374E2B7FD2D vdev_path=/dev/disk/by-id/wwn-0x50014ee262071313-part3
Mar 11 22:44:06 hubble zed[21035]: eid=9395 class=vdev.bad_ashift pool_guid=0x38BF9374E2B7FD2D vdev_path=/dev/disk/by-id/wwn-0x50014ee20cb16a07-part3
Mar 11 22:44:07 hubble zed[21053]: eid=9396 class=history_event pool_guid=0x38BF9374E2B7FD2D
Mar 11 22:44:07 hubble zed[21057]: eid=9397 class=resilver_start pool_guid=0x38BF9374E2B7FD2D
Mar 11 22:44:07 hubble zed[21062]: eid=9398 class=history_event pool_guid=0x38BF9374E2B7FD2D
Mar 11 22:44:18 hubble zed[21754]: eid=9399 class=config_sync pool_guid=0x38BF9374E2B7FD2D
Mar 11 22:44:21 hubble zed[22114]: eid=9400 class=config_sync pool_guid=0x38BF9374E2B7FD2D
Mar 11 22:44:21 hubble zed[22190]: eid=9401 class=vdev.bad_ashift pool_guid=0x38BF9374E2B7FD2D vdev_path=/dev/disk/by-id/wwn-0x50014ee20cb16a07-part3
Mar 11 22:44:22 hubble zed[22201]: eid=9402 class=vdev.bad_ashift pool_guid=0x38BF9374E2B7FD2D vdev_path=/dev/disk/by-id/wwn-0x50014ee262071313-part3
Mar 11 22:44:22 hubble zed[22204]: eid=9403 class=history_event pool_guid=0x38BF9374E2B7FD2D
Mar 11 22:44:22 hubble zed[22208]: eid=9404 class=resilver_start pool_guid=0x38BF9374E2B7FD2D
Mar 11 22:44:22 hubble zed[22210]: eid=9405 class=history_event pool_guid=0x38BF9374E2B7FD2D
Mar 11 22:44:36 hubble zed[23236]: eid=9406 class=config_sync pool_guid=0x38BF9374E2B7FD2D
Mar 11 22:44:36 hubble zed[23309]: eid=9407 class=config_sync pool_guid=0x38BF9374E2B7FD2D
Mar 11 22:44:42 hubble zed[23562]: eid=9408 class=vdev.bad_ashift pool_guid=0x38BF9374E2B7FD2D vdev_path=/dev/disk/by-id/wwn-0x50014ee262071313-part3
Mar 11 22:44:42 hubble zed[23569]: eid=9409 class=vdev.bad_ashift pool_guid=0x38BF9374E2B7FD2D vdev_path=/dev/disk/by-id/wwn-0x50014ee20cb16a07-part3
Mar 11 22:44:43 hubble zed[23581]: eid=9410 class=history_event pool_guid=0x38BF9374E2B7FD2D
Mar 11 22:44:43 hubble zed[23584]: eid=9411 class=resilver_start pool_guid=0x38BF9374E2B7FD2D
Mar 11 22:44:43 hubble zed[23588]: eid=9412 class=history_event pool_guid=0x38BF9374E2B7FD2D
```

`systemctl stop zfs-zed` solved the problem, and resilvering is continuing normally (I hope).

**Comment by 4oo4** [comment]
@as-com Did stopping `zfs-zed` end up solving it? I tried that without any luck, still got stuck in the same resilver loop.

However, what seemed to fix it for me was doing a `detach` on the troublesome disk, zeroing it out with `scrub -p fillzero /dev/sda` (NOT zfs' `scrub`), and then re-attaching it. Haven't gotten an error yet, and the resilver seems to be going about 30% faster than before.

I think what's happening (at least in my case) is that there's some kind of I/O error that zfs can't recover from during the resilver that triggers the restart, but forcing it to do a clean resilver (since I'm guessing it will try to be lazy and avoid writing blocks that are already there) finally corrects it.

**Comment by as-com** [comment]
@4oo4 Yes, stopping zed solved the issue. My hypothesis is that it's a timing thing where zed sometimes requests configuration stuff from zfs which causes the resilver to restart.

**Comment by 4oo4** [comment]
@as-com Spoke too soon, it restarted again. Last night I tried disabling that service but still had it restart.

However, it's going even faster than before (no errors after that either) since I stopped it again. `zfs-zed` was definitely triggering the resilver, having a clean disk to work with seems to help too.

On zfs version 0.7.8.

**Comment by StefanCohen** [comment]
I have reproduced my problem multiple times by now (no errors. just a continuously restart of the resilvering process) and stoping zed during the resilvering always helps for me.

**Comment by 4oo4** [comment]
@StefanCohen Cool, so I'm guessing it's OK to restart zed once the resilvering is done?

**Comment by StefanCohen** [comment]
@4oo4 Yes. That is what I do.

**Comment by DigitalBites** [comment]
This saved my bacon! systemctl stop zfs-zed was the winner. If this is standard practice, the docs should be updated. I had all kinds of trouble with resilver restarting as I was proactively replacing a drive due to increased smart errors and couldn't figure out why they wouldn't progress. There were no read or write errors so was very confusing until I read this.

Thank you!!

**Comment by sco01** [comment]
This bug has been open since 2012 so I agree that it would be nice if the docs mention it.

**Comment by yanfali** [comment]
I just ran into this while replacing some older drivers with newer ones. Thanks for having this issue open. In the end it was a combination of TLER and ZED that allowed my storage to make forward progress.

**Comment by wrouesnel** [comment]
I've just been running into this for the last few days on an array. Killing zed and letting it try again.

**Comment by WanWizard** [comment]
This issue saved my bacon. 

Got stuck in the resilver loop for days on end, with a constant fear of a failure of the remaining disk in the mirror vdev. Stopped zfs-zed, disk reslivered in minutes.

**Comment by wrouesnel** [comment]
For me I needed to do 2 things - kill zfs-zed and also set `zfs_scan_ignore_errors=1` to let my fairly messed up array resilver itself.

Ultimately it looks like I'll need to extract all the remaining data and rebuild from backups though.

**Comment by chrisrd** [comment]
Hey all, original poster (from 7 years ago!) here...

I was running into this "resilver constantly restarting" after replacing a drive, which naturally kicked off a resilver.

One of the symptoms was these messages spewing into my `zpool history -i` every 15 seconds or so:

```
2019-04-01.10:52:54 [txg:564945133] scan aborted, restarting errors=0
2019-04-01.10:52:54 [txg:564945133] starting deferred resilver errors=0
2019-04-01.10:52:54 [txg:564945133] scan setup func=2 mintxg=3 maxtxg=564899810
```

This is with 0.8.0, and WITHOUT the zfs feature `feature@resilver_defer` enabled.

As soon as I enabled `feature@resilver_defer` the restarting stopped occurring and the resilver began making progress.

For others in this situation the options seem to be:

1. (temporarily?) revert to 0.7.latest and let the resilver finish
2. enable just `feature@resilver_defer`:
```
zpool set feature@resilver_defer=enabled pool`
```
3. enable all 0.8.0 features:
```
zpool upgrade pool
```

**WARNING** enabling any 0.8.0 features, including `feature@resilver_defer`, will result in not being able to revert to 0.7.x with this pool.

(This is a completely different cause to the original problem from 7 years ago, but the symptoms are the same and this issue seems to be the place people come when they hit this "resilver constantly restarting".)

**Comment by ptx0** [comment]
yes I just hit the same issue this morning with a grub boot pool (features disabled, mostly) that I attached a new disk to a 9-way mirror, it resilvered 1.5G repeatedly for more than 2 days, until exporting and reimporting the pool, after which the final resilver finished. I can't enable the resilver_defer feature, it is a grub pool. this bug is pretty annoying.

**Comment by JoakimZiegler** [comment]
For what it's worth, I also just hit this issue today, after having upgraded to 0.8.1 with a faulted drive, because with 0.7 the resilver was going to take a very long time, so I wanted sequential resilver. Stopping zed didn't work, but upgrading the pool and enabling resilver_defer solved the problem.

**Comment by Lalufu** [comment]
I'm currently experiencing this on a pool running under 0.8.2, with feature@resilver_defer enabled (the pool was on 0.8.2, with that feature enabled, when the resilver started). The original disk is no longer present, the replacement disk encountered errors during the first resilver run (which I cleared), and finished the second resilver run without any errors. It's currently in the third resilver.

**Comment by Lalufu** [comment]
```
  pool: tank
 state: DEGRADED
status: One or more devices is currently being resilvered.  The pool will
        continue to function, possibly in a degraded state.
action: Wait for the resilver to complete.
  scan: resilver in progress since Mon Jan 20 14:53:56 2020
        590G scanned at 497M/s, 229G issued at 193M/s, 14.4T total
        32.2G resilvered, 1.55% done, 0 days 21:27:01 to go
config:

        NAME                                            STATE     READ WRITE CKSUM
        tank                                            DEGRADED     0     0     0
          raidz2-0                                      DEGRADED     0     0     0
            ata-WDC_WD30EFRX-68AX9N0_WD-WCC1T1489518    ONLINE       0     0     0
            ata-WDC_WD30EFRX-68AX9N0_WD-WCC1T1489979    ONLINE       0     0     0
            ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N7PV8232    ONLINE       0     0     0
            replacing-3                                 DEGRADED     0     0     0
              9593279170455378102                       UNAVAIL      0     0     0  was /dev/disk/by-id/ata-WDC_WD30EFRX-68AX9N0_WD-WCC1T1545830-part1
              ata-WDC_WD60EFAX-68SHWN0_WD-WX11D39929YU  ONLINE       0     0     0  (resilvering)
            ata-WDC_WD30EFRX-68AX9N0_WD-WCC1T1545960    ONLINE       0     0     0
            ata-WDC_WD30EFRX-68AX9N0_WD-WCC1T1550817    ONLINE       0     0     0
        logs    
          mirror-1                                      ONLINE       0     0     0
            wwn-0x55cd2e404b510699-part6                ONLINE       0     0     0
            wwn-0x55cd2e404b511223-part6                ONLINE       0     0     0



NAME  PROPERTY                       VALUE                          SOURCE
tank  size                           16.2T                          -
tank  capacity                       88%                            -
tank  altroot                        -                              default
tank  health                         DEGRADED                       -
tank  guid                           13759943618182453813           -
tank  version                        -                              default
tank  bootfs                         -                              default
tank  delegation                     on                             default
tank  autoreplace                    off                            default
tank  cachefile                      -                              default
tank  failmode                       wait                           default
tank  listsnapshots                  off                            default
tank  autoexpand                     off                            default
tank  dedupditto                     0                              default
tank  dedupratio                     1.00x                          -
tank  free                           1.83T                          -
tank  allocated                      14.4T                          -
tank  readonly                       off                            -
tank  ashift                         12                             local
tank  comment                        -                              default
tank  expandsize                     -                              -
tank  freeing                        0                              -
tank  fragmentation                  38%                            -
tank  leaked                         0                              -
tank  multihost                      off                            default
tank  checkpoint                     -                              -
tank  load_guid                      11289215483194079640           -
tank  autotrim                       off                            default
tank  feature@async_destroy          enabled                        local
tank  feature@empty_bpobj            active                         local
tank  feature@lz4_compress           active                         local
tank  feature@multi_vdev_crash_dump  enabled                        local
tank  feature@spacemap_histogram     active                         local
tank  feature@enabled_txg            active                         local
tank  feature@hole_birth             active                         local
tank  feature@extensible_dataset     active                         local
tank  feature@embedded_data          active                         local
tank  feature@bookmarks              enabled                        local
tank  feature@filesystem_limits      enabled                        local
tank  feature@large_blocks           active                         local
tank  feature@large_dnode            enabled                        local
tank  feature@sha512                 enabled                        local
tank  feature@skein                  enabled                        local
tank  feature@edonr                  enabled                        local
tank  feature@userobj_accounting     active                         local
tank  feature@encryption             enabled                        local
tank  feature@project_quota          active                         local
tank  feature@device_removal         enabled                        local
tank  feature@obsolete_counts        enabled                        local
tank  feature@zpool_checkpoint       enabled                        local
tank  feature@spacemap_v2            active                         local
tank  feature@allocation_classes     enabled                        local
tank  feature@resilver_defer         enabled                        local
tank  feature@bookmark_v2            enabled                        local
```

**Comment by thisisnotmyrealname** [comment]
This is still going on in 0.8.3

  pool: tv
 state: DEGRADED
status: One or more devices is currently being resilvered.  The pool will
        continue to function, possibly in a degraded state.
action: Wait for the resilver to complete.
  scan: resilver in progress since Thu Mar 26 15:16:14 2020
        138G scanned at 17.3G/s, 52.7G issued at 6.59G/s, 81.6T total
        0B resilvered, 0.06% done, 0 days 03:31:11 to go
config:

        NAME                        STATE     READ WRITE CKSUM
        tv                          DEGRADED     0     0     0
          raidz3-0                  DEGRADED     0     0     0
            wwn-0x5000cca261c9c282  ONLINE       0     0     0
            wwn-0x5000cca261c9bc3d  OFFLINE      0     0     0
            wwn-0x5000c500b2e78f6e  ONLINE       0     0     0
            wwn-0x5000cca261c9a642  ONLINE       0     0     0
            wwn-0x5000cca261c9079e  OFFLINE      0     0 74.0K
            wwn-0x5000c500b5b16651  ONLINE       0     0     0
            wwn-0x5000c500b2e71d11  ONLINE       0     0     0
            wwn-0x5000c500b5b15fdc  ONLINE       0     0     0
            wwn-0x5000c500b5b15d85  ONLINE       0     0     0
            wwn-0x5000c500b2e67d21  ONLINE       0     0     0
            wwn-0x5000c500b4e0ac49  ONLINE       0     0     0
            wwn-0x5000c500b5b05687  ONLINE       0     0     0
            wwn-0x5000cca261c9bdc0  ONLINE       0     0     0
            wwn-0x5000cca261c9ad25  ONLINE       0     0     0
            wwn-0x5000cca261c9ac0c  ONLINE       0     0     0
            wwn-0x5000cca261c8662b  ONLINE       0     0     0
        cache
          wwn-0x5002538500204a4f    ONLINE       0     0     0

errors: No known data errors

I offlined the disks and stopped ZED to see if that will fix the resilver, which apparently has been running non-stop for >15 days.

**Comment by rootless4real** [comment]
I just found this thread.  Thank you to everyone who has posted over the years.  I'm running into problems with ZFS suffering errors and degrading disks in any zpool I create or run through my CK23601.  It doesn't matter if the controller is an LSI 2008 or an LSI 2308, the same problem is occurring.  I do happen to have a single WD EADS disk.  I will remove it and see what impact it has.  I've tried using all SAS or all SATA disks in the same zpool without success.  The entire setup is: (2x) Norco 4224 to (2x) CK23601 to a single LSI 2308.  The LSI 2308 (LSI 9206-16e) is hosted in a Dell 910.   Any insight would be appreciated.  Thank you.

**Comment by rootless4real** [comment]
I believe it was the EADS drive.  I pulled it out.  At the same time I also moved all of the SAS into one enclosure and the SATA into another enclosure.  

The piece I still haven't been able to resolve - the ZFS filesystem I created on SAS disks connected directly to the controller generates errors when connected via SAS-expander. 

Thanks again for the above discussion.

**Comment by bar-g** [comment]
The archlinux wiki (https://wiki.archlinux.org/title/ZFS#Pool_resilvering_stuck.2Frestarting.2Fslow.3F) hinted about a workaround for failing/restarting resilver operations caused by problematic/buggy behavior of the "ZFS event daemon", and references this issue.

However, from looking here it's not-obvious whether this long standing resilvering issue has a workaournd, can be improved/re-created or even solved.
* the relevant comments are hidden (collapsed) by default
* this issue may rather **not** be a documentation issue, but something assignable-to / fixable-in "zfs event daemon"?

**Comment by bar-g** [comment]
Summary of comments, why this going on does not seem a only-doc-issue (to not use SAS expanders to connect SATA disks):
* stoping zfs event daemon (zed) during resilver allowed the resilver to finish
https://github.com/openzfs/zfs/issues/840#issuecomment-386430330 (repeatedly confirmed later)

And there were hints that the **related root cause** for this can be worked around by
* adjusting TLER/ERC/CCTL drive error timeout settings or
* doing disk writes:
https://github.com/openzfs/zfs/issues/840#issuecomment-440409812
https://github.com/openzfs/zfs/issues/840#issuecomment-450959054
